keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,993 @@
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Most batch jobs consist of multiple (usually two) roles of tasks. For example, a
+typical Tensorflow job has some parameter servers and workers. Each role
+requires a quorum to be ready to run. With that being considered, the following
+API designs all employ a 2-tier specification (Subsets) inside PodGroup. 
+
+The PodGroup API will be defined as a top-level API (in group scheduling.k8s.io, like PriorityClass) 
+
+```go
+// group scheduling.k8s.io
+type PodGroup struct {
+	Spec   PodGroupSpec
+	Status PodGroupStatus
+}
+
+// PodGroupSpec is a description of a PodGroupSpec.
+type PodGroupSpec struct {
+	// Subsets consist of various kinds of pod sets.
+	// A PodGroup is schedulable if all subsets can be scheduled.
+	Subsets []Subset
+
+	// ScheduleTimeoutSeconds defines the timeout threshold to abort
+	// an in-progress PodGroup-level scheduling attempt.
+	ScheduleTimeoutSeconds *int32
+}
+
+// Subset represents a collection of pods with the same role.
+type Subset struct {
+	// Name is used to distinguish pods with different roles in the
+	// same pod group. Optional if there is only one subset in the 
+	// pod group.
+	Name string
+
+	// MinMember specifies the minimum number of pods required for a 
+	// subset to be operational.
+	MinMember int32
+}
+```
+
+In order to associate a pod with its target PodGroup, we need to add a field call `PodGroup` in the podSpec to declare which PodGroup and Subset the pod belongs to.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+  ....
+  PodGroup PodGroupSpec
+	....
+}
+
+// - pod.spec.podGroup.name = foo
+// - pod.spec.podGroup.subset = (bar/baz)
+type PodGroupSpec struct {
+  // Name is the name of PodGroup.
+  Name   string
+  // Subset is the name of PodGroup.
+  Subset string
+}
+```
+
+In this design, a PodGroupStatus is needed to record the latest status. This
+enables users to quickly know what’s going on with this PodGroup; also some
+integrators (like ClusterAutoscaler) can rely on it to make pending PodGroup
```

Comment:
You probably need to clarify when the status will be updated with respect to the Pod unschedulable condition (is this going to be updated if the PodGroup is not satisfied yet)?

Reply:
also what happens if one of the pods gets deleted after the group was scheduled? There are likely cases where a pod (or more) getting deleted can be considered ok, and other cases where the whole group should be preempted.

Reply:
> You probably need to clarify when the status will be updated with respect to the Pod unschedulable condition (is this going to be updated if the PodGroup is not satisfied yet)?

Added in the scheduling part of implementation.

> also what happens if one of the pods gets deleted after the group was scheduled? There are likely cases where a pod (or more) getting deleted can be considered ok, and other cases where the whole group should be preempted.

According to the current use case, if one (or more) pods is deleted, the results are determined by the workload controller whether to continue or rebuild or mark a failure.